src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+        self.try_add_capabilities(memory_model.required_capabilities());
+
+        instruction.add_operand(addressing_model as u32);
+        instruction.add_operand(memory_model as u32);
+        instruction
+    }
+
+    fn instruction_entry_point(
+        &mut self,
+        entry_point: &crate::EntryPoint,
+        ir_module: &crate::Module,
+    ) -> Instruction {
+        let mut instruction = Instruction::new(Op::EntryPoint);
+        let function_id = self
+            .lookup_function
+            .lookup_id(&entry_point.function)
+            .unwrap();
+
+        instruction.add_operand(entry_point.exec_model as u32);
+        instruction.add_operand(function_id);
+
+        if self.debug_enabled {
+            let mut debug_instruction = Instruction::new(Op::Name);
+            debug_instruction.set_result(function_id);
+            debug_instruction.add_operands(self.string_to_words(entry_point.name.as_str()));
+            self.debugs.push(debug_instruction);
+        }
+
+        instruction.add_operands(self.string_to_words(entry_point.name.as_str()));
+
+        let function = &ir_module.functions[entry_point.function];
+        for ((handle, _), &usage) in ir_module
+            .global_variables
+            .iter()
+            .zip(&function.global_usage)
+        {
+            if usage.contains(crate::GlobalUse::STORE) || usage.contains(crate::GlobalUse::LOAD) {
+                let id = self.get_global_variable_id(
+                    &ir_module.types,
+                    &ir_module.global_variables,
+                    &handle,
+                );
+                instruction.add_operand(id);
+            }
+        }
+
+        self.try_add_capabilities(entry_point.exec_model.required_capabilities());
+        match entry_point.exec_model {
+            ExecutionModel::Vertex | ExecutionModel::GLCompute => {}
+            ExecutionModel::Fragment => {
+                let execution_mode = ExecutionMode::OriginUpperLeft;
+                self.try_add_capabilities(execution_mode.required_capabilities());
+                let mut execution_mode_instruction = Instruction::new(Op::ExecutionMode);
+                execution_mode_instruction.add_operand(function_id);
+                execution_mode_instruction.add_operand(execution_mode as u32);
+                for word in execution_mode_instruction.to_words() {
+                    self.module.logical_layout.execution_modes.push(word)
+                }
+            }
+
+            _ => unimplemented!("{:?}", entry_point.exec_model),
+        }
+
+        instruction
+    }
+
+    fn get_type_id(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        handle: &crate::Handle<crate::Type>,
+    ) -> Word {
+        match self.lookup_type.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let (instruction, id) = self.instruction_type_declaration(arena, *handle);
```

Comment:
shouldn't we add it to the lookup table?

Reply:
It is inserted, but not at this place. It is getting inserted in the `instruction_type_declaration` function, but I completely agree that it will be far better to insert here, for both constantly having to think about inserting and for duplicate code.